server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -777,6 +784,11 @@ void innerOnResponse(SearchResponse searchResponse) {
             SearchResponse createFinalResponse() {
                 return searchResponseMerger.getMergedResponse(clusters);
             }
+
+            @Override
+            protected void releaseResponse(SearchResponse searchResponse) {
+                searchResponse.decRef();
```

Comment:
Why do we incRef and decRef the SearchResponse in SearchResponseMerger, but in the ActionListener we only decRef? (My guess is that the Transport layer does the incRef before passing it to ActionListeners - is that correct?)

And since there is a chain of ActionListeners here, how do we know which one to decRef in? (Is that ActionListener.releaseAfter does?)

Reply:
> how do we know which one to decRef in? 

See my other answer (https://github.com/elastic/elasticsearch/pull/103021#discussion_r1417402650). We don't need to know and listeners must make sure to be reference count neutral from the perspective of the caller.
The important thing to note really is that we always start at reference count 1 when creating an instance.